### ISSUE TITLE
Option to Use Classes Instead of Style

### ISSUE THREAD

**Issue by mpql** [chetachiezikeuzor/Highlightr-Plugin#11]
Hey there! First, I wanted to say that this is a neat plugin idea. Thanks for writing it, and for sharing it with the community!

Currently, if I highlight some text, `example text`, and pick, say, `blue`, I get the following:

```html
<mark style="background: #ADCCFFA6;">example text</mark>
```

If I were to change the value of "Blue" (which, an "edit" option doesn't seem to currently be available in the plugin's settings near as I can tell, but I can delete and recreate that name, for instance), if the file is open, I think it removes any instances of the highlight, and if it's not, then the highlight's mark remains configured to the old color. This makes it so notes can become "outdated" relative to a given configuration, which I think is not really compatible with the idea of a knowledge database.

Is there anything I'm missing, like some practical reason why classes cannot or should not be used for this?

My proposed solution would instead have it so your interface curates its own list of classes, while the plugin adds classes to notes, which can be updated / changed over time to reflect the user's current use of / view on highlights, e.g.

So the markdown could look something like:
```html
<mark class="highlightr-blue">example text</mark>
```

and elsewhere, something like:
`highlightr-user.css`
```css
.highlightr-blue {
  background: #ADCCFFA6;
}
```

Thanks again for sharing your plugin!

EDIT: An option to cost classes versus inline styling would likely be appropriate to cover all use cases.

**Comment by chetachiezikeuzor** [comment]
> You've got to be being disingenuous. Option paralysis for this? Please. If you knew what you were talking about vis a vis UX, the settings wouldn't currently be laid out so haphazardly. But if you insist, you can take out the option to have styling flattened into user markdowns, and just use classes by default.
> 
> In any case, considering you've failed to address any of my above points over repeated posts, I will assume you have nothing salient to say.

@mpql 
Looking back at this conversation, I have to say that this was a pretty rude statement to make towards my friend Ryan. Considering the fact that he was behind the data restructuring of the plugin, I feel that his input here was judged unfairly. Calling him disingenuous is far from true, especially since what he is addressing **is** something the user will inevitably have to face. Now I do think that the feature has value, so I have implemented it. But I ask that we keep these discussions respectful. We all have our own opinions, but there's no need to cast judgments towards each other.

**Comment by mpql** [comment]
> @mpql Looking back at this conversation, I have to say that this was a pretty rude statement to make towards my friend Ryan. Considering the fact that he was behind the data restructuring of the plugin, I feel that his input here was judged unfairly. Calling him disingenuous is far from true, especially since what he is addressing **is** something the user will inevitably have to face. Now I do think that the feature has value, so I have implemented it. But I ask that we keep these discussions respectful. We all have our own opinions, but there's no need to cast judgments towards each other.

Hi @chetachiezikeuzor. First off, I want to say I appreciate your work. No conversation can be had without first acknowledging that. I appreciate both the work you've put into this plugin, and into my request in specific. Thank you.

All said, I disagree that I was being unfairly rude, be he friend, contributor, or otherwise. I found Ryan to be combative and condescending from his second comment in this thread. Eventually, I said the quiet part out loud, but I reject that I was the source of that; I did bring it to the forefront, but I don't regret that.

If you care to hear my point of view here, I repeatedly explained my perspective in-depth, and he would ignore the bulk of my points and make a very short comment with a new reason to exclude my points, again moving the goalposts. It went from being about renderers, to data portability and vendor lock-ins, to the ideals of Obsidian, to the existence of an option causing choice paralysis and code complexity, hurting long-term reliability. And that **is** disingenuous.

Choice paralysis matters for adoption; I would imagine a free, open-source plugin cares more about being useful than market share, but even if I suspend that assumption, choice paralysis is easily mitigated with clean and clear options pages (which this did not formerly have, thus [my other issue](https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/12)), sectioned options (such as Advanced Settings, or hidden settings available in a `.json` or similar), clear, simple options, and good defaults. At the time of my posting, the options page for this had two options, and a list for colors, which does not begin to approach choice paralysis territory. Now, I don't expect everyone to know all of that, but the only reason I can think of for invoking a concept one demonstrably doesn't understand is in the hopes I wouldn't be able to contest it. Making room for both sets of behavior doesn't seem unreasonable to me, and I feel as though I was not approached in a collaborative way from the outset.

This is, I feel, a problem with the culture of the open-source community at large. Not just that two people can have differing opinions, but that people feel they have the right to decide that another person's perspective is incorrect based on their own, and shouldn't exist at all. You don't owe me labor. Ryan doesn't owe me labor. Either or both of you could say "I would not find this useful, and it doesn't fit with my perspective for what this tool does, or what I think users will find useful, so I'm not gonna work on it." And I'd fork the thing myself and submit a PR when I've got some time. Repeatedly returning to post to put down an idea for ever-changing reasoning from a cursory perspective is _not constructive_. The only conceivable goal I can see there is to have my issue closed rather than considered by other parties, and I think using any position he has with you to do that is unfair.

If you've read this far, thanks. I appreciate your perspective, and all said, I don't intend to engage with Ryan any further, so you don't need to worry about my future behavior. I do think this was important and worth addressing, however, because I respect the work you've contributed, and don't want you to think I was being flippant in my censure.

I mean this sincerely: have a good day, and thanks again. :)

**Comment by ryanpcmcquen** [comment]
I don't see any goal post movement here, just responses to your arguments for this feature. I stand behind all the points made here, and genuinely believe them to be true.

**Comment by mpql** [comment]
> I don't see any goal post movement here, just responses to your arguments for this feature. I stand behind all the points made here, and genuinely believe them to be true.

Okay, but do you also genuinely believe you're being constructive by replying here and now? Or in your previous posts? I'm not sure I'd take the assurance of someone whose argument validity I contest that their arguments are valid. They'd be fine against it being the default -- which I relented immediately upon your first comment, and both commented to that effect, and edited my original post -- but not against it being a possible option. Your arguments above seek to end the suggestion and its collaboration, and they speak to your competence in the same fields inversely to your authenticity in making them, so I will leave you to decide where that line is. As I've said, I don't wish to engage with you any further, so please don't engage with me.

**Comment by taneltm** [comment]
So much passion in this thread!

I was initially going to suggest wrapping the inline-style based option in a custom element. But after reading this thread, I'm not sure about my idea anymore. I'll share it anyway.

User writes: `<mark-red>test</mark-red>` or maybe `<m-red>test</m-red>`    
Custom element renders:
```html
<mark style="background: red">test</mark>
or
<mark style="text-decoration: underline; text-decoration-color: red;">test</mark>
based on plugin settings
```
This would also allow the user to play around with the plugin settings and figure our how they want their highlights to look like without committing to a specific style when they insert the highlights. I guess that's also a benefit for the class based option.

But now that I'm thinking about it, is it even possible to use HTML custom elements in Markdown because of security reasons or maybe in Obsidian it wouldn't have to be an actual HTML custom element?

This `<mark-red>` or `<m-red>` option would be easier to type for the user, but wouldn't be compatible with other Markdown renderers - they would just ignore those tags, but still render the text inside them.  
I think then there needs to be a command which transforms the custom element tags in the current document to regular `<mark>` tags with inline styles, so the document can be easily exported for other renderers - This might be a good idea for the class-based option as well.

**Comment by ryanpcmcquen** [comment]
> > I don't see any goal post movement here, just responses to your arguments for this feature. I stand behind all the points made here, and genuinely believe them to be true.
> 
> Okay, but do you also genuinely believe you're being constructive by replying here and now? Or in your previous posts? I'm not sure I'd take the assurance of someone whose argument validity I contest that their arguments are valid. They'd be fine against it being the default -- which I relented immediately upon your first comment, and both commented to that effect, and edited my original post -- but not against it being a possible option. Your arguments above seek to end the suggestion and its collaboration, and they speak to your competence in the same fields inversely to your authenticity in making them, so I will leave you to decide where that line is. As I've said, I don't wish to engage with you any further, so please don't engage with me.

Yes, my comments here are constructive. I see no condescending tone. Just factual responses targeting the fallibility of the suggestion. You have decided to read into my tone for some reason, and failed to offer responses to many of the caveats here.

**Comment by mpql** [comment]
> So much passion in this thread!
> 
> I was initially going to suggest wrapping the inline-style based option in a custom element. But after reading this thread, I'm not sure about my idea anymore. I'll share it anyway.
> 
> User writes: `<mark-red>test</mark-red>` or maybe `<m-red>test</m-red>` Custom element renders:
> 
> ```
> <mark style="background: red">test</mark>
> or
> <mark style="text-decoration: underline; text-decoration-color: red;">test</mark>
> based on plugin settings
> ```
> 
> This would also allow the user to play around with the plugin settings and figure our how they want their highlights to look like without committing to a specific style when they insert the highlights. I guess that's also a benefit for the class based option.
> 
> But now that I'm thinking about it, is it even possible to use HTML custom elements in Markdown because of security reasons or maybe in Obsidian it wouldn't have to be an actual HTML custom element?
> 
> This `<mark-red>` or `<m-red>` option would be easier to type for the user, but wouldn't be compatible with other Markdown renderers - they would just ignore those tags, but still render the text inside them. I think then there needs to be a command which transforms the custom element tags in the current document to regular `<mark>` tags with inline styles, so the document can be easily exported for other renderers - This might be a good idea for the class-based option as well.

This is actually a great idea, but I think the implementation might need to be styled element around custom element for portability. The custom element could be styled by classes, and since it's the inner element, those will take precedence over the style attribute, and for those who don't use it, they will have the styled `<mark>` take precedence over the unstyled element, with the option to style the custom element later. It could, however, also just be a known inner element, e.g. `<span style="#FFFF00"><span class="highlightr-yellow">Text to be highlighted</span></span>` -- the core idea here to me is wrapping for both data portability and forward-compatibility if anyone ever wants to style their classes inside obsidian.

@ryanpcmcquen I don't believe I'm doing those things, and you certainly haven't been answering my points for need, or my ideas for fixing things, so I don't believe your comments were coming from a collaborative place, regardless of any tone I might "read into" your comments from, say, unnecessarily-placed quotation marks. But go ahead and show me where I've failed to reply, and I will answer them (if I haven't already), and do the same for you.

**Comment by ryanpcmcquen** [comment]
It's typical in the english language to use quotation marks when quoting other people, it helps with context.

I am yet to see a response on how copy and paste is handled. And how the code complexity affects the codebase for all the people who don't want this feature.

**Comment by mpql** [comment]
> It's typical in the english language to use quotation marks when quoting other people, it helps with context.

Dude, this way of speaking has been the issue from the beginning. I tried to ask in genuinely good faith what you believed unaddressed issues were, but there you go again with the condescension. I dare you to explain how you improved clarity here -- you certainly didn't improve mine. Imagine reading:

> Well, there are use cases outside of 
> > converters
>
> . What about good old fashioned copy and paste? Do you remember to copy external files when working with plain text?

Wouldn't it seem so weird? It's plain to see that clarity wasn't your goal; you sought to make me appear foolish because you didn't like my suggestion. And I don't think trying to intellectually bully people -- even if I _were_ wrong -- is the right way to approach that -- this kind of attitude is _poison_ to open source. Just own your behavior, dude. And if you are genuinely not trying to be a jerk right now, I would seriously reevaluate how you speak to people, because either your motives or your methods are just awful, and I'm left with the very distinct impression that I'm feeding a troll. I'd wager this is far from the first time you've been reacted to this way, whichever case is true.

> I am yet to see a response on how copy and paste is handled.

You've seen two, in fact; I just don't think you're trying very hard. Before you even made that argument, I suggested my feature request be an option. Copy+paste continues to work for anyone who doesn't change the default behavior. taneltm above suggested nested elements, which, if using my suggested implementation:
```html
<span style="color:#F00"><span class="highlightr-red">Highlighted</span></span>
```
...would also continue to work via copy+paste.

> And how the code complexity affects the codebase for all the people who don't want this feature.

Preponderance of evidence lies with the claimant. You have yet to establish how a toggle option significantly increases code complexity or "long term reliability" (I'm quoting it because I think it's disingenuous hokum) by any appreciable margin, considering the save buttons don't even work. Since you also seemed to think three options approaches choice paralysis (it doesn't), I'll point out that code complexity refers to exponential interactions, which isn't relevant here at all. Complicated != complex in that context, but it's neither.

I think you're trying to find technicalities you think make you correct so you can ignore that you're speaking to other humans. If someone says something you don't like, they must be "wrong" for some reason or another. You've got to keep reaching until something sticks. Either way, the issue has been resolved, and there's nothing to be gained from continuing this. Closing this issue. Later. ✌

**Comment by ryanpcmcquen** [comment]
I would argue that the most poisonous thing to open source is inundating a project with feature requests without actually contributing any code.

I'm yet to see anything that actually handles copy/paste, only the concession that it is now an option to enable breaking behavior.

The code complexity is now at its worst case, since now both scenarios have to be handled. This increases the complexity of the regex, and hurts overall performance. Any future requests or changes need to accommodate two possible syntaxes. That's detrimental at an architectural level to this plugin. Ockham's razor applies here.

There has also been zero discussion on how switching between the two syntaxes is handled, if at all.

Random question, if you were building a programming language, would you implement two syntaxes for the same result?

**Comment by akadormouse** [comment]
> would you implement two syntaxes for the same result?
Not wanting to get involved in a private disagreement, but feel I have to point out that this is what markdown does.

**Comment by taneltm** [comment]
> Random question, if you were building a programming language, would you implement two syntaxes for the same result?

Not from the start, but I would add syntactic sugar later when I see a need for it and I think this applies here.
```html
<mark style="background: #ADCCFFA6;">example text</mark>
<mark class="highlightr-blue">example text</mark>
<mark-blue">example text</mark-blue>
```
They do achieve the same result, but the syntactic sugar of the more compact and human-readable variants sacrifice compatibility.

JavaScript has syntactic sugar which was added later as well:
- Function+bind vs. arrow function
- Promises vs. async-await
- `strA.indexOf(strB) != -1` vs. `strA.contains(strB)`
- ...
But syntactic sugar doesn't have to make the initial option obsolete. You still need the Promise API in some cases and and a `function` has the benefit of being hoisted and without having automatic context binding.
And of course the syntactic sugar in JS sacrifices compatibility as well.

**Comment by ryanpcmcquen** [comment]
I'm not a fan of syntactic sugar. I've seen these kind of things create unforeseen problems at scale. There's a fundamental thing we probably disagree on here, I don't consider JavaScript to be a well designed language, so providing examples of it demonstrating this behavior doesn't sway my experience or opinion.

If you've been with JavaScript for a while, and read things like Crockford's 'JavaScript: The Good Parts', a large portion of that book is dedicated to advocating for only a subset of the language to be used, as a best practice.

**Comment by taneltm** [comment]
> There's a fundamental thing we probably disagree on here, I don't consider JavaScript to be a well designed language, so providing examples of it demonstrating this behavior doesn't sway my experience or opinion.

I agree with you on that, that JS is badly designed... if it even ever was "designed". I picked JS because the extension is written in that (well TS) and I thought you'd understand the examples. I think this syntactic sugar is what makes JavaScript less crappy, but you don't have to agree with me.

Anyway. I only came here to suggest the custom elements idea. If you don't like the idea or if you don't see it happening in this project, that's fine.
